### PR TITLE
Fix split bug

### PR DESCRIPTION
**Pull request by c4pt0r**

### CODE REVIEW

File: scan.go [hunk 1]
```diff
@@ -235,7 +261,7 @@ func (s *Scan) processResponse(response pb.Message) []*ResultRow {
 	}
 
 	if n < s.numCached {
-		s.nextStartRow = incrementByteString(s.location.EndKey, len(s.location.EndKey)-1)
+		s.nextStartRow = s.location.EndKey //incrementByteString(s.location.EndKey, len(s.location.EndKey)-1)
```
Comment (c4pt0r) [inline scan.go]: comment style

File: scan_test.go [hunk 2]
```diff
@@ -87,7 +88,7 @@ func (s *ScanTestSuit) TestScan(c *C) {
 				}
 				cnt++
 			}
-			c.Assert(cnt, Equals, 4)
+			c.Assert(cnt, Equals, 4445)
```
Comment (c4pt0r) [inline scan_test.go]: This test seems out of date.

File: scan.go [hunk 3]
```diff
@@ -18,6 +17,19 @@ func incrementByteString(d []byte, i int) []byte {
 		return append(make([]byte, 1), r...)
 	}
 	r[i]++
+	for i > 0 {
```
Comment (c4pt0r) [inline scan.go]: add some comments, or example.

**Comment by morefreeze** [comment]
@c4pt0r I have fix missing data when region splitting
PTAL

File: scan_test.go (hunk 2 above)
Comment (morefreeze) [inline scan_test.go]: So I updated it

File: scan.go (hunk 1 above)
Comment (morefreeze) [inline scan.go]: yep, I would delete it in future commit

**Comment by c4pt0r** [comment]
👌

**Comment by ngaut** [comment]
Well done. Please do some clean up.

File: scan.go [hunk 4]
```diff
@@ -165,9 +181,10 @@ func (s *Scan) getData(nextStart []byte) []*ResultRow {
 	if s.id > 0 {
 		req.ScannerId = pb.Uint64(s.id)
 	}
-	if s.StartRow != nil {
-		req.Scan.StartRow = s.StartRow
-	}
+	req.Scan.StartRow = nextStart
+	//if s.StartRow != nil {
```
Comment (c4pt0r) [inline scan.go]: remove usless codes

File: scan.go [hunk 5]
```diff
@@ -234,22 +252,23 @@ func (s *Scan) processResponse(response pb.Message) []*ResultRow {
...
 		s.id = 0
-		lastRegionRows = 0
 	}
 
 	if n == 0 && !nextRegion {
 		s.Close()
 	}
 
+	if s.skipFirst {
+		results = results[1:len(results)]
```
Comment (c4pt0r) [inline scan.go]: [1:]

File: client.go [hunk 6]
```diff
@@ -38,6 +38,7 @@ const (
 	retryLongerMultiplier    = 31
 	socketDefaultRetryWaitMs = 200
 	defaultRetryWaitMs       = 100
+	nines                    = "99999999999999"
```
Comment (c4pt0r) [inline client.go]: Add a comment here

File: scan_test.go [hunk 7]
```diff
@@ -50,6 +50,7 @@ func (s *ScanTestSuit) TearDownSuite(c *C) {
 
 func (s *ScanTestSuit) TestScanInSplit(c *C) {
 	for {
+		break
```
Comment (c4pt0r) [inline scan_test.go]: seems strange here

File: client.go [hunk 8]
```diff
@@ -38,6 +38,8 @@ const (
 	retryLongerMultiplier    = 31
 	socketDefaultRetryWaitMs = 200
 	defaultRetryWaitMs       = 100
+	// always >= any unix timestamp(hbase version)
+	nines = "99999999999999"
```
Comment (ngaut) [inline client.go]: s/nines/beyondMaxHbaseTimestamp        ?

File: scan.go [hunk 9]
```diff
@@ -264,7 +280,11 @@ func (s *Scan) nextBatch() int {
 		startRow = s.StartRow
 	}
 	rs := s.getData(startRow)
-	if rs == nil || len(rs) == 0 {
+	// current region get 0 data, switch next region
+	if rs != nil && len(rs) == 0 && s.nextStartRow != nil {
```
Comment (ngaut) [inline scan.go]: s/nextStartRow/nextStartRowKey  ?

Comment (morefreeze) [inline scan.go]: ok, I'll check all variable name about row or key. They seem confused.

File: scan_test.go (hunk 7 above)
Comment (morefreeze) [inline scan_test.go]: Oooops, I'll change it to normal for loop

File: client.go (hunk 8 above)
Comment (morefreeze) [inline client.go]: Hmmmm, `beyondMaxTimestamp` seems enough

**Pull request merged**